mlir/lib/Dialect/Rock/Transforms/RockMultibuffer.cpp

```diff
@@ -0,0 +1,452 @@
+    auto upperDims = llvm::to_vector(t.getUpperDims());
+    for (auto &d : lowerDims) {
+      d++;
+    }
+    for (auto &d : upperDims) {
+      d++;
+    }
+    newOps.push_back(TransformAttr::get(ctx, t.getType(), t.getParams(),
+                                        t.getUpperNames(), upperDims,
+                                        t.getLowerNames(), lowerDims));
+  }
+
+  // Create and add the passthrough transform. Notice that logically
+  // every loop will have its own multibuffer. We already added
+  // a broadcast transformation on top of the transform stack
+  // to route the indices back to the real (smaller) multibuffer
+  SmallVector<int64_t, 4> multiBufferedUpperBounds{loopLength};
+  SmallVector<int64_t, 4> multiBufferedLowerBounds{loopLength};
+  ArrayRef<int64_t> originalUpperBounds = transform.getUpperBounds();
+  ArrayRef<int64_t> originalLowerBounds = transform.getLowerBounds();
+  llvm::append_range(multiBufferedUpperBounds, originalUpperBounds);
+  llvm::append_range(multiBufferedLowerBounds, originalLowerBounds);
+  auto passThrough =
+      TransformAttr::get(ctx, TransformType::PassThrough, {},
+                         {"multi_buffer_idx"}, {0}, {"multi_buffer_idx"}, {0});
+  newOps.push_back(passThrough);
+  TransformMapAttr newMap = TransformMapAttr::get(
+      newOps, multiBufferedUpperBounds, multiBufferedLowerBounds);
+  return newMap;
+}
+
+/// Helper function to create a memref.subview to extract the multibuffer
+/// slice.
+Value getSubviewIntoMultibuffer(RewriterBase &rewriter, Location loc, Value val,
+                                Value index, ArrayRef<int64_t> originalShape,
+                                MemRefType mbType) {
+  int64_t mbMemRefTypeRank = mbType.getRank();
+  IntegerAttr zero = rewriter.getIndexAttr(0);
+  IntegerAttr one = rewriter.getIndexAttr(1);
+  SmallVector<OpFoldResult> offsets(mbMemRefTypeRank, zero);
+  SmallVector<OpFoldResult> sizes(mbMemRefTypeRank, one);
+  SmallVector<OpFoldResult> strides(mbMemRefTypeRank, one);
+  // Offset is [bufferIndex, 0 ... 0 ].
+  offsets.front() = index;
+  // Sizes is [1, original_size_0 ... original_size_n ].
+  for (int64_t i = 0, e = originalShape.size(); i != e; ++i)
+    sizes[1 + i] = rewriter.getIndexAttr(originalShape[i]);
+  // Strides is [1, 1 ... 1 ].
+  auto dstMemref =
+      cast<MemRefType>(memref::SubViewOp::inferRankReducedResultType(
+          originalShape, mbType, offsets, sizes, strides));
+  Value subview = rewriter.create<memref::SubViewOp>(loc, dstMemref, val,
+                                                     offsets, sizes, strides);
+  return subview;
+}
+
+/// This is replacing a generic operation op(buffer) with a sequence of
+/// - multibuffer_view = memref.subview(buffer)
+/// - op(multibuffer_view)
+Operation *subviewAndClone(RewriterBase &rewriter, Location loc, Operation *op,
+                           Value val, Value loopIndex, int64_t operandNumber,
```

Comment:
If I understood your comment correctly you are saying that I can pass simply the `OpOperand` instead of `{operation*, operandNumber}`, right? This is what I did, but if I misunderstood the comment, please let me know